hazelcast/src/main/java/com/hazelcast/config/PersistentMemoryConfig.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Configuration class for persistent memory devices (e.g. Intel Optane).
+ */
+public class PersistentMemoryConfig {
+    /**
+     * Paths to the non-volatile memory directory.
+     */
+    private List<PersistentMemoryDirectoryConfig> directoryConfigs = new LinkedList<>();
+
+    public PersistentMemoryConfig() {
+    }
+
+    /**
+     * Constructs an instance with copying the fields of the provided
+     * {@link PersistentMemoryConfig}.
+     *
+     * @param persistentMemoryConfig The configuration to copy
+     */
+    public PersistentMemoryConfig(PersistentMemoryConfig persistentMemoryConfig) {
+        persistentMemoryConfig.directoryConfigs
+                .forEach(directoryConfig -> addDirectoryConfig(new PersistentMemoryDirectoryConfig(directoryConfig)));
+    }
+
+    /**
+     * Returns the persistent memory directory configurations to be used
+     * to store memory structures allocated by native memory manager.
+     * <p>
+     * By default there are no configuration is set indicating that
+     * volatile RAM is being used.
+     */
+    public List<PersistentMemoryDirectoryConfig> getDirectoryConfigs() {
+        return directoryConfigs;
+    }
+
+    /**
+     * Sets the persistent memory directory configuration to the set of
+     * directories provided in the {@code directoryConfigs} argument.
+     * <p/>
+     * If the specified directories are not unique either in the directories
+     * themselves or in the NUMA nodes specified for them,
+     * {@link InvalidConfigurationException} is thrown. Setting the NUMA
+     * node on the subset of the configured directories while leaving
+     * not set on others also results in {@link InvalidConfigurationException}.
+     *
+     * @param directoryConfigs The persistent memory directories to set
+     * @return this {@link PersistentMemoryConfig} instance
+     * @throws InvalidConfigurationException If the configured directories
+     *                                       violate consistency or
+     *                                       uniqueness checks.
+     */
+    public PersistentMemoryConfig setDirectoryConfigs(List<PersistentMemoryDirectoryConfig> directoryConfigs) {
```

Comment:
Can you sprinkle some nullable and non-null annotations on the method parameters and return values?

Reply:
Done.